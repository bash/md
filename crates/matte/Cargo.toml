[package]
name = "matte"
version = "0.1.0"
edition = "2021"

[dependencies]
anstyle.workspace = true
bat.workspace = true
fmtastic.workspace = true
hostname.workspace = true
matte-language-names = { path = "../matte-language-names" }
pulldown-cmark.workspace = true
smallvec.workspace = true
trait-set.workspace = true
unicode-linebreak-chunked.workspace = true
unicode-width.workspace = true
url.workspace = true

[dev-dependencies]
insta = { workspace = true, features = ["glob"] }
serde = { workspace = true, features = ["derive"] }
toml.workspace = true
