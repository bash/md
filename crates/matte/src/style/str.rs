use crate::style::StyleExt as _;
use anstyle::{Reset, Style};
use pulldown_cmark::CowStr;
use std::borrow::Borrow as _;
use std::fmt;
use std::ops::Deref;

#[derive(Debug)]
pub(crate) struct StyledStr<'a>(pub(crate) CowStr<'a>, pub(crate) Style);

impl<'a> StyledStr<'a> {
    pub(crate) fn on_top_of(self, fallback: Style) -> Self {
        Self(self.0, self.1.on_top_of(fallback))
    }
}

impl<'a> StyledStr<'a> {
    pub(crate) fn new(s: impl Into<CowStr<'a>>, style: Style) -> Self {
        Self(s.into(), style)
    }
}

impl StyledStr<'_> {
    pub(crate) fn borrowed(&self) -> StyledStr<'_> {
        StyledStr(CowStr::Borrowed(self.0.borrow()), self.1)
    }
}

impl Default for StyledStr<'_> {
    fn default() -> Self {
        Self::new("", Default::default())
    }
}

impl Deref for StyledStr<'_> {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'a> From<&'a str> for StyledStr<'a> {
    fn from(value: &'a str) -> Self {
        Self::new(value, Style::default())
    }
}

impl<'a> From<String> for StyledStr<'a> {
    fn from(value: String) -> Self {
        Self::new(value, Style::default())
    }
}

impl<'a> From<CowStr<'a>> for StyledStr<'a> {
    fn from(value: CowStr<'a>) -> Self {
        Self(value, Style::default())
    }
}

impl fmt::Display for StyledStr<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.1.is_plain() {
            write!(f, "{}", self.0)
        } else {
            write!(f, "{}{}{}", self.1, self.0, Reset)
        }
    }
}
