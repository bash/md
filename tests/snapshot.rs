use insta::{assert_snapshot, glob};
use matte::{render, supported_parser_options, HeadingDecoration, Options};
use pulldown_cmark::{Event, MetadataBlockKind, Options as ParserOptions, Parser, Tag};
use serde::{Deserialize, Deserializer};
use std::fs::read_to_string;
use std::str::FromStr;

#[test]
fn test_snippets() {
    glob!("snippets/*.md", |path| {
        let markdown = read_to_string(path).unwrap();
        let rendered = render_to_string(&markdown);
        assert_snapshot!(rendered);
    })
}

fn render_to_string(input: &str) -> String {
    let parser = Parser::new_ext(input, supported_parser_options());
    let mut buffer = Vec::new();
    let snippet_options = read_snippet_options(input);
    let mut options = Options::plain_text(snippet_options.width.unwrap_or(120));
    options.heading_decoration = Some(
        snippet_options
            .heading_decoration
            .map(|h| h.0)
            .unwrap_or(HeadingDecoration::default()),
    );
    render(parser, &mut buffer, options).unwrap();
    String::from_utf8(buffer).unwrap()
}

fn read_snippet_options(input: &str) -> SnippetOptions {
    toml::de::from_str(&extract_frontmatter(input)).unwrap()
}

fn extract_frontmatter(input: &str) -> String {
    let parser = Parser::new_ext(
        input,
        ParserOptions::ENABLE_PLUSES_DELIMITED_METADATA_BLOCKS,
    );
    let mut frontmatter = String::new();
    let mut inside_frontmatter = false;
    for event in parser {
        match event {
            Event::Start(Tag::MetadataBlock(MetadataBlockKind::PlusesStyle)) => {
                inside_frontmatter = true
            }
            Event::Text(text) if inside_frontmatter => frontmatter.push_str(&text),
            _ => break,
        }
    }
    frontmatter
}

#[derive(Deserialize)]
struct SnippetOptions {
    width: Option<u16>,
    heading_decoration: Option<DeserializeFromStr<HeadingDecoration>>,
}

struct DeserializeFromStr<T>(T);

impl<'de, T> serde::Deserialize<'de> for DeserializeFromStr<T>
where
    T: FromStr,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let str: String = Deserialize::deserialize(deserializer)?;
        T::from_str(&str)
            .map(DeserializeFromStr)
            .map_err(|_| serde::de::Error::custom("invalid value"))
    }
}
