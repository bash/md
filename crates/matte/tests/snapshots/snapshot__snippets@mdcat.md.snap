---
source: tests/snapshot.rs
expression: rendered
input_file: tests/snippets/mdcat.md
---
[1m[34m┈[0m[1m[34mCommonMark[0m

[1m[34m┈┈[0m[1m[34mMarkup[0m

[3m[33mmdcat[0m supports standard [3minline[0m [1mmarkup[0m and block markup, even with syntax highlighting for code 
blocks.

[3m    [0m[3mDon't believe everything you read on the Internet.[0m
[3m    [0m
[3m    [0m[3m—Abraham Lincoln[0m

[35mfn[0m [34mmain[0m() {
    println!([32m"[0m[32mHello world![0m[32m"[0m);
}

[1m[34m┈┈[0m[1m[34mSpecials[0m

[3m[33mmdcat[0m prints images and hyperlinks:

[7m🖼 Rust[0m

Source: ]8;id=1;https://www.rust-lang.org/\Rust]8;;\

Enjoy ❤️
